pkg/authorization/api.go

```diff
@@ -0,0 +1,46 @@
+package authorization
+
+const (
+	// POST method for daemon request authorization
+	AuthZApiRequest = "AuthZPlugin.AuthZReq"
+
+	// POST method for daemon response authorization
+	AuthZApiResponse = "AuthZPlugin.AuthZRes"
+
+	// Name of the interface all AuthZ plugins implement
+	AuthZApiImplements = "authz"
+)
+
+// Request holds data required for authZ plugins
+type Request struct {
+	// User holds the user extracted by AuthN mechanism
+	User string `json:"User,omitempty"`
+
+	// UserAuthNMethod holds the mechanism used to extract user details (e.g., krb)
+	UserAuthNMethod string `json:"UserAuthNMethod,omitempty"`
+
+	// RequestMethod holds the HTTP method (GET/POST/PUT)
+	RequestMethod string `json:"RequestMethod,omitempty"`
+
+	// RequestUri holds the full HTTP uri (e.g., /v1.21/version)
+	RequestUri string `json:"RequestUri,omitempty"`
+
+	// RawRequest stores the raw request sent to the docker daemon
+	RawRequest []byte `json:"RawRequest,omitempty"`
+
+	// RawRequest stores the raw response sent from docker daemon
+	RawResponse []byte `json:"RawResponse,omitempty"`
+}
+
+// Request is the expect object that returns from the authZ plugin
+type Response struct {
```

Comment:
@NathanMcCauley, thanks for you comments.
Re /logs and /attache, the authentication framework should only be called on the HTTP requests. Once the http connection is hijacked, no additional authorization should to be perform. This pattern enables full authorization capabilities to docker as described in the current requirements (for framework and specific RBAC plugin).  The only change in current code is to replace the response recorder with concrete object that implements hijack.

Re response rewrites, since the api version is exposed to the plugin, I think it should be the plugin responsibility to determine whether it can or cannot modify the response or which parts can be changed.  As far as I know, the docker daemon api is backward compatible, so plugins are essentially equivalent to regular client communicating with the api. 

I totally agree with other comments, please let me know your thoughts on the above items, and I will modify the code appropriately.